lesson 5/OpenCVtest/tensor.cpp

```diff
@@ -0,0 +1,504 @@
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		delete[] data[i];
+	}
+	delete[] data;
+}
+
+Tensor2D::Tensor2D(const Tensor2D& copyObj) {
+	shape = copyObj.shape;
+	allocate2D(shape[0], shape[1]);
+
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		std::copy(&copyObj.data[i][0], &copyObj.data[i][shape[1]], &data[i][0]);
+	}
+}
+
+Tensor2D::Tensor2D(Tensor2D&& moveObj) : Tensor2D() {
+	swap(*this, moveObj);
+}
+
+void swap(Tensor2D& left, Tensor2D& right) {
+	std::swap(left.data, right.data);
+	std::swap(left.shape, right.shape);
+}
+
+void Tensor2D::printElements() const {
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		for (std::size_t k = 0; k < shape[1]; ++k) {
+			std::cout << data[i][k] << " " << std::flush;
+		}
+		std::cout << std::endl;
+	}
+}
+
+Tensor2D& Tensor2D::operator=(const Tensor2D& copyAss) {
+	Tensor2D tmp(copyAss);
+	swap(*this, tmp);
+	return *this;
+}
+
+Tensor2D& Tensor2D::operator=(Tensor2D&& moveAss) {
+	Tensor2D tmp(std::move(moveAss));
+	swap(*this, tmp);
+	return *this;
+}
+
+std::array<std::size_t, 2> Tensor2D::getShape() const {
+	return shape;
+}
+
+void Tensor2D::transpose() {
+	Tensor2D tmp(shape[1], shape[0]);
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		for (std::size_t k = 0; k < shape[1]; ++k) {
+			tmp.data[k][i] = data[i][k];
+		}
+	}
+	swap(*this, tmp);
+}
+
+std::vector<double> Tensor2D::flatten() const {
+	std::vector<double> flattend(shape[0] * shape[1]);
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		std::copy(&data[i][0], &data[i][shape[1]], flattend.begin() + i * shape[1]);
+	}
+	return flattend;
+}
+
+void Tensor2D::setValue(int row, int colum, double value) {
+	if (row >= shape[0] || row < 0 || colum >= shape[1] || colum < 0) {
+		throw WrongSize();
+	}
+		data[row][colum] = value;
+}
+
+double& Tensor2D::getValue(std::size_t row, std::size_t column ){
+	if ( row >= shape[0] || row < 0 || column >= shape[1] || column < 0) {
+		throw WrongSize();
+	}
+	return data[row][column];
+}
+
+Tensor2D& Tensor2D::matmul(const Tensor2D& right) {
+	if (shape[1] != right.shape[0]) {
+		throw WrongSize();
+	}
+	std::vector<double> tmp;
+	
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		for (std::size_t j = 0; j < right.shape[1]; ++j) {
+			double sum = 0;
+			for (std::size_t k = 0; k < right.shape[0]; ++k) {
+				sum += data[i][k] * right.data[k][j];
+			}
+			tmp.push_back(sum);
+		}
+	}
+	Tensor2D tmpTens(shape[0], right.shape[1]);
+	int help = 0;
+	for (std::size_t i = 0; i < shape[0]; ++i) {
+		//std::copy(tmp.begin() + i * right.shape[1], tmp.end() - (right.shape[1]*(right.shape[1]-i+1 )),std::back_inserter(tmpTens.data[i][0]));
+		for (std::size_t k = 0; k < shape[0]; ++k) {
+			tmpTens.data[i][k] = tmp[help++];
+		}
+	}
+	shape[1] = right.shape[1];
+	swap(*this, tmpTens);
+	return *this;
+}
+
+Tensor2D::Tensor2D(const cv::Mat& mat) {
+	const int channels = mat.channels();
+	if (channels == 1) {
+		shape = { (std::size_t)mat.rows, (std::size_t)mat.cols };
+		allocate2D(mat.rows, mat.cols);
```

Comment:
instead call your newly overloaded version: 
`if (mat.isContinuous()) {
			allocate2D(mat.rows, mat.cols, mat.begin<uchar>());
		} `
to avoid double traversal